Add onboarding Playground flow

zaerl opened the pull request:
## Proposed Changes

* Install WordPress Playground dependencies
* Load Playground on the `onboarding/playground` step
  * If a Blueprint is provided in the URL hash, it will be used to load Playground
  * Playground is loaded with OPFS to preserve the site after a reload
* Submit the `onboarding/playground` step to `onboarding/domains` and maintain the `playground` id in the query string
* If a `playground` query string is present, after checkout redirect to `site-setup/importPlayground`
* Add a `site-setup/importPlayground` step that loads Playground based on the provided `playground` query string
  * After loading Playground, the site is exported into a ZIP
  * Based on the `siteId` and `siteSlug` query parameters the ZIP is posted to the `/imports/new` API endpoint
  * After the ZIP is posted, the page is redirected to a existing `site-setup/importerWordpress` step

Side note: I needed to change a not-related unit test to let the build work. Bumped as well Odyssey limits.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* We want to enable users to start a new WordPress.com site from Playground.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Test only in Chrome** (OPFS doesn't work correctly in Firefox)

* [Start the Playground flow](http://calypso.localhost:3000/setup/onboarding/playground)
* Make some changes to your site (update the title, install a plugin...)
* Click _Launch on WordPress.com_
* Choose a domain
* Select the _Business plan_
* Checkout
* You should be redirected to `/setup/site-setup/importerPlayground`
* A site upload should start
* You should be redirected to `/setup/site-setup/importerWordpress` after the upload
* The import should start running
  * If the import fails it's ok for this PR, but please report your error in a comment, so that we can investigate it further

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

zaerl committed:
Add onboarding/playground feature flag

zaerl committed:
Add new Playground step

zaerl committed:
Add new Playground step

bgrgicak commented:
@zaerl [here's my PR with Playground loaded using the JS API](https://github.com/Automattic/wp-calypso/pull/100769).

It supports loading Blueprints from the hash just like the Playground website and it uses OPFS (so don't test it in Firefox).

My next step is to connect the _Launch on WordPress.com_ button to the import API.

bgrgicak committed:
Load Playground using the JS API in onboarding (#100769)

* Load Playground from the JS API

* Add playground to the onboarding flow

* Redirect to Playground after checkout

* Basic setup screen

* Add import flow TODO add API requests

* Limit Playground setup to logged in users

* Run import using the WP.com API

* Use import flow to import Playground zip

* fix redirect

* Fix WordPress import UI

* Add option=content to importerWordpress url

* cleanup

bgrgicak committed:
Load Playground using the JS API in onboarding (#100769)

* Load Playground from the JS API

* Add playground to the onboarding flow

* Redirect to Playground after checkout

* Basic setup screen

* Add import flow TODO add API requests

* Limit Playground setup to logged in users

* Run import using the WP.com API

* Use import flow to import Playground zip

* fix redirect

* Fix WordPress import UI

* Add option=content to importerWordpress url

* cleanup

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/components/playground-iframe/index.tsx:
```diff
@@ -0,0 +1,32 @@
+			return;
+		}
+
+		if ( playgroundClient ) {
```
I assume you meant 
```suggestion
		if ( ! playgroundClient ) {
```

zaerl replied:
This is the proper check. In this way, `initializeWordPressPlayground` is called only if the client hasn't already initialized it using `startPlaygroundWeb`. Do you also think this is the correct reasoning?

alshakero replied:
Oh you are right. I just pattern-detected an early return against a nullish value.

alshakero commented on client/landing/stepper/declarative-flow/onboarding.ts:
```diff
@@ -195,6 +198,18 @@ const onboarding: Flow = {
+			const playgroundId = getQueryArg( window.location.href, 'playground' );
+			if ( playgroundId && providedDependencies.siteSlug ) {
+				return [
+					addQueryArgs( withLocale( '/setup/site-setup/importerPlayground', locale ), {
```
I don't think you need the locale since the user is logged in at this point.

zaerl replied:
This must be used? [http://calypso.localhost:3000/setup/onboarding/playground](https://href.li/?http://calypso.localhost:3000/setup/onboarding/playground) can be accessed logged-in and not.

alshakero commented on client/landing/stepper/declarative-flow/onboarding.ts:
```diff
@@ -373,16 +388,25 @@ const onboarding: Flow = {
+						 */
+						const playgroundId = getQueryArg( window.location.href, 'playground' );
+						const redirectTo: string = playgroundId
+							? addQueryArgs( withLocale( '/setup/site-setup/importerPlayground', locale ), {
```
+1

alshakero commented on client/landing/stepper/declarative-flow/onboarding.ts:
```diff
@@ -373,16 +388,25 @@ const onboarding: Flow = {
+									siteId: providedDependencies.siteId,
+									playground: playgroundId,
+							  } )
+							: addQueryArgs( withLocale( '/setup/onboarding/post-checkout-onboarding', locale ), {
```
+1

alshakero reviewed (commented):
I'm very excited about this. I love it. I left some comments meant to help, not to block you or anything. 

As a side note, I recommend leaving `onboarding` untouched and instead copying a simple flow like `example.ts` and make your own new flow, for your own sanity. 

When you make a separate flow, you can just deploy without worrying about flags and breaking things.

ashfame committed:
Ashfame/playground wpcom signup flow (#101257)

* add todo

* remove unnecessary await

* disable lint rule

* fix vertical height of playground's iframe

* switch playground packages to private repo src

* update playground packages to 1.0.29

* use uuid as playground id to reduce collision probability to practically zero

* Update style.scss

Co-authored-by: Bero <berislav.grgicak@gmail.com>

---------

Co-authored-by: Bero <berislav.grgicak@gmail.com>

ashfame committed:
Playground onboarding predefined blueprints (#101647)

* add OPFS_PATH_PREFIX constant

* abstract figuring out blueprint from url to its own func

* predefine blueprints which are used based on blueprint GET param

* only run default blueprint post on subsequent runs

* rely on isWordPressInstalled instead of firstRun to specify which blueprint should run

zaerl committed:
Update yarn.lock

zaerl committed:
Add missing Step types

ashfame committed:
define some predefined blueprints

ashfame committed:
add support for resolving blueprints from 'blueprint-url' GET param (#101789)

zaerl committed:
Remove unused flow declaration

zaerl committed:
Add flag check

zaerl committed:
Update yarn.lock

ashfame committed:
Use PHP version that's in use for new WoA sites (#101826)

* use usePhpVersions().recommendedValue instead of hardcoding a specific version, this will keep it in sync with what WoA sites use to spin new sites

* encapsulate blueprint management in a closure, so we are always overwriting the preferred php version

zaerl committed:
Fix: wrong checks

zaerl committed:
Update Jest snapshot

zaerl committed:
Add missing NL

zaerl committed:
Remove debug code

zaerl committed:
Fix Odyssey stats size limit

zaerl committed:
Update yarn lock

ashfame commented:
Couple of things:

1) When I load the following page directly (skipping the checkout):

```
http://calypso.localhost:3000/setup/site-setup/importerPlayground?siteSlug=ashfametest.wpcomstaging.com&siteId=242499700&playground=b1f59d6a-ce7b-4c9d-8481-ab075bb52930
```

Business plan on my site is not detected and I am still presented the screen to choose a business plan on:

```
http://calypso.localhost:3000/setup/site-setup/importerWordpress?siteSlug=ashfametest.wpcomstaging.com&siteId=242499700&playground=b1f59d6a-ce7b-4c9d-8481-ab075bb52930&option=content
```

I also had trouble associating the business plan to this particular site after going through the checkout too. So, it's most likely a specific/localised issue to my test site, so should be ok to ignore if your test site functions as intended.

2) `< Back` hyperlink (adjacent to logo in top right corner), on domains selection page takes you back to sites page and not the playground page. So essentially, that page is lost unless you type the url or use the browser's back button.

I think, we might want to think if the back button shown on different steps as per our onboarding flow makes sense: Does it link to where it should? Is that consistent with browser's back button? Does it make sense to hide it on certain screens?

ashfame reviewed (approved):
LGTM!

alshakero commented on client/landing/stepper/declarative-flow/flows/onboarding/onboarding.ts:
```diff
@@ -195,6 +203,20 @@ const onboarding: Flow = {
+				: null;
+			if ( playgroundId && providedDependencies.siteSlug ) {
+				return [
+					addQueryArgs( withLocale( '/setup/site-setup/importerPlayground', locale ), {
+						siteSlug: providedDependencies.siteSlug,
+						siteId: providedDependencies.siteId,
+						playground: playgroundId,
+					} ),
```
We don't need locale in site-setup flow because the user is logged and the param will be ignored.

alshakero commented on client/landing/stepper/declarative-flow/flows/onboarding/onboarding.ts:
```diff
@@ -373,16 +397,25 @@ const onboarding: Flow = {
+						 */
+						const playgroundId = getQueryArg( window.location.href, 'playground' );
+						const redirectTo: string = playgroundId
+							? addQueryArgs( withLocale( '/setup/site-setup/importerPlayground', locale ), {
```
No need for the locale.

alshakero commented on client/landing/stepper/declarative-flow/flows/onboarding/onboarding.ts:
```diff
@@ -373,16 +397,25 @@ const onboarding: Flow = {
+									siteId: providedDependencies.siteId,
+									playground: playgroundId,
+							  } )
+							: addQueryArgs( withLocale( '/setup/onboarding/post-checkout-onboarding', locale ), {
```
Same here.

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/components/playground-setup/index.tsx:
```diff
@@ -0,0 +1,82 @@
+			<DocumentHead title={ __( 'Playground Setup' ) } />
+			<StepWrapper
+				hideBack={ shouldHideBackButton() }
+				hideSkip={ shouldHideSkip() }
```
Since these simply return true,

```suggestion
				hideBack
				hideSkip
```

zaerl replied:
Done. Thanks.

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/components/playground-setup/index.tsx:
```diff
@@ -0,0 +1,82 @@
+
+	// There's probably a better way to do this.
+	const siteSlug = getQueryArg( window.location.href, 'siteSlug' )?.toString();
+	const siteId = parseInt( getQueryArg( window.location.href, 'siteId' )?.toString() ?? '0' );
```
Yes, you can use [`useSiteData`](https://github.com/Automattic/wp-calypso/blob/63f61027743517b1192d9a9b268c2f8e764e70da/client/landing/stepper/declarative-flow/flows/entrepreneur-flow/entrepreneur-flow.ts#L60) hook. 

```
const { siteId, siteSlug } = useSiteData();
```

zaerl replied:
👍

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/components/playground-setup/index.tsx:
```diff
@@ -0,0 +1,82 @@
+	const isPlaygroundEligible = useIsPlaygroundEligible();
+	if ( ! isPlaygroundEligible ) {
+		window.location.assign( '/start' );
+	}
```
I would move this to the bottom and do 
```suggestion
	if ( ! isPlaygroundEligible ) {
		return window.location.assign( '/start' );
	}
```

This would prevent the component from rendering for a second before redirecting.

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/index.tsx:
```diff
@@ -0,0 +1,69 @@
+
+	const shouldHideSkip = () => {
+		return true;
+	};
```
I think these are noise.

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/index.tsx:
```diff
@@ -0,0 +1,69 @@
+			<DocumentHead title={ __( 'Playground' ) } />
+			<StepWrapper
+				hideBack={ shouldHideBackButton() }
+				hideSkip={ shouldHideSkip() }
```
```suggestion
				hideBack
				hideSkip
```

zaerl replied:
Done!

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/lib/import-playground.ts:
```diff
@@ -0,0 +1,29 @@
+import { zipWpContent, type PlaygroundClient } from '@wp-playground/client';
+import { uploadExportFile } from 'calypso/state/imports/actions';
+
+export async function getSiteZip( playground: PlaygroundClient ): Promise< File > {
```
The type is inferrable by TS, I'd remove it to cut down on noise and future maintenance. 

```suggestion
export async function getSiteZip( playground: PlaygroundClient ) {
```

zaerl replied:
Make sense.

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/lib/initialize-playground.ts:
```diff
@@ -0,0 +1,195 @@
+						resource: 'wordpress.org/themes',
+						slug: 'twentytwentyfour',
+					},
+					options: {
+						activate: true,
+					},
+				},
+			],
+		},
+		2023: {
+			...DEFAULT_BLUEPRINT,
+			steps: [
+				{
+					step: 'installTheme',
+					themeData: {
+						resource: 'wordpress.org/themes',
+						slug: 'twentytwentythree',
+					},
+					options: {
+						activate: true,
+					},
+				},
+			],
+		},
+		design1: {
+			...DEFAULT_BLUEPRINT,
+			steps: [
+				{
+					step: 'installTheme',
+					themeData: {
+						resource: 'wordpress.org/themes',
+						slug: 'variations',
+					},
+					options: {
+						activate: true,
+					},
+				},
+			],
+		},
+	};
```
It seems all of these are static information that can be outside the function.

zaerl replied:
Moved.

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/lib/initialize-playground.ts:
```diff
@@ -0,0 +1,195 @@
+	const url = new URL( window.location.href );
+	let playgroundId = url.searchParams.get( 'playground' );
+	if ( ! playgroundId ) {
+		playgroundId = crypto.randomUUID();
```
Damn that's sweet!

zaerl replied:
👍

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/lib/initialize-playground.ts:
```diff
@@ -0,0 +1,195 @@
+import config from '@automattic/calypso-config';
+import { Blueprint } from '@wp-playground/blueprints';
+import { MountDescriptor, PlaygroundClient, startPlaygroundWeb } from '@wp-playground/client';
```
Nit: 
```suggestion
import { type MountDescriptor, typePlaygroundClient, startPlaygroundWeb } from '@wp-playground/client';
```

alshakero reviewed (changes requested):
Did a thorough review because this touches our main income source. This is pretty sweet and exciting!

ashfame committed:
fix back link on domains screen

bgrgicak commented on client/landing/stepper/declarative-flow/internals/steps-repository/playground/lib/initialize-playground.ts:
```diff
@@ -0,0 +1,195 @@
+const blueprintManager = ( () => {
+	const DEFAULT_BLUEPRINT: Blueprint = {
+		preferredVersions: {
+			php: '8.3',
```
Should we use `usePhpVersions().recommendedValue` here instead of hardcoding 8.3?

ashfame replied:
It's always overwritten with the specified recommended php version value. Had a comment on that line which got removed. Reason why I wrapped both constants in its closure, was to prevent the direct usage. I guess we can remove it altogether or make it act as a fallback when an invalid value for recommendedPhpVersion is supplied.

ashfame replied:
Made it more obvious - https://github.com/Automattic/wp-calypso/pull/100715/commits/688c24b23b150123f41355fa571091f1fc26c257

zaerl committed:
Remove useless functions

zaerl committed:
Remove type returns

zaerl committed:
Move constants outside the function

zaerl committed:
Move to useSiteData

ashfame committed:
move blueprints to its own file simplifying stuff

zaerl committed:
Fix: remove hardcoded date

gabrielcaires committed:
Update tests to make then wait until the conditions are ready

zaerl committed:
Update yarn.lock

zaerl committed:
Fix: update snapshot

zaerl committed:
Run yarn dedupe

zaerl committed:
Remove spurious spaces

zaerl committed:
Fix: return before rendering again

zaerl committed:
Fix: return before rendering again

zaerl committed:
Fix: wrong types

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17318195

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Hi @zaerl, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Playground Setup`
- `Launch on WordPress.com`

Thank you in advance!

tyxla commented:
It looks like this was merged with @alshakero's blocking review - can we confirm that all feedback was addressed before merging?

zaerl commented:
> It looks like this was merged with @alshakero's blocking review - can we confirm that all feedback was addressed before merging?

Hi, yes. It was.

https://github.com/Automattic/wp-calypso/pull/100715#discussion_r2013689018
https://github.com/Automattic/wp-calypso/pull/100715#discussion_r2013694707
Removed here https://github.com/Automattic/wp-calypso/pull/100715/commits/7e7cc1760be056239f022017c5ef558a923221c7

https://github.com/Automattic/wp-calypso/pull/100715#discussion_r2013694707
Fixed here https://github.com/Automattic/wp-calypso/pull/100715/commits/914eea3d05d3586f914e27655fd360f086626757

https://github.com/Automattic/wp-calypso/pull/100715#discussion_r2013813060
Moved here https://github.com/Automattic/wp-calypso/pull/100715/commits/1a6b00596143b665730290ad7816a819045cdc51

The only not removed code is `withLocale` call, because `/setup/onboarding/playground` can [be accessed](https://github.com/Automattic/wp-calypso/pull/100715#discussion_r1985341304) by logged-in and not logger-in users.
